### PR TITLE
[CELEBORN-834][DOC] Add fault tolerant document

### PR DESCRIPTION
**Pull request by waitinfuture**
### What changes were proposed in this pull request?
As title.

### Why are the changes needed?
As title.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Manual test.

### CODE REVIEW

**Comment by waitinfuture** [comment]
cc @pan3793 @FMX

File: docs/developers/faulttolerant.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+- For each data batch, `ShuffleClient` adds a `(Map Id, Attempt Id, Batch Id)` header, in which
+  `Batch Id` is a unique id for the data batch in the map attempt
+- `LifecycleManager` keeps all `PartitionLocation`s with the same partition id
+- For each `PartitionLocation` split, at least one replica is successfully committed before shuffle read
+- `LifecycleManager` records the successful task attempt for each map id, and only data from that attempt is read
+  for the map id
+- `ShuffleClient` discards data batches with a batch id that it has already read
+
+
```
Comment (AngersZhuuuu) [inline docs/developers/faulttolerant.md:83]: Maybe we can also add some content about 
1. timeout checker
2. Blacklist
3. network rack aware
4.

File: docs/developers/faulttolerant.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+In addition to data replication to handle `Worker` lost, Celeborn tries to handle exceptions during shuffle
+as much as possible, especially the following:
+
+- When `PushData`/`PushMergedData` fail
+- When fetch chunk fails
+- When disk is unhealthy or reaching limit
+
+This article is based on [ReducePartition](../../developers/storage#reducepartition).
+
+## Handle PushData Fail
```
[the remark this answers is not in the document]
Comment (waitinfuture) [reply docs/developers/faulttolerant.md:30]: It's similar, I'll add more details on this.

File: docs/developers/faulttolerant.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
+---
+license: |
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+  http://www.apache.org/licenses/LICENSE-2.0
```
Comment (pan3793) [inline docs/developers/faulttolerant.md:9]: Emm... this line format is broken... in many docs files.

File: docs/developers/faulttolerant.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+Celeborn considers no `DataLost` IFF every `PartitionLocation` has succeeded to commit at least one replica (if
+replication is turned off, there is only one replica for each `PartitionLocation`).
+
+When a `Worker` is down, all `PartitionLocation`s on the `Worker` will be revived, causing `Revive` RPC flood
+to `LifecycleManager`. To alleviate this, `ShuffleClient` batches all `Revive` requests before sending to
+`LifecycleManager`:
+
+![BatchRevive](../../assets/img/batchrevive.svg)
+
+## Handle Fetch Fail
```
Comment (pan3793) [inline docs/developers/faulttolerant.md:50]:
```suggestion
## Handle Fetch Failure
```

File: docs/developers/faulttolerant.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+`ShuffleClient` defines the max number of retries for each replica(defaults to 3). When fetch chunk fails,
+`ShuffleClient` will try another replica (in case where replication is off, retry the same one).
+
+If the max retry number exceeds, `ShuffleClient` gives up retrying and throws Exception.
+
+## Disk Check
+`Worker` periodically checks disk health and usage. When health check fails, `Worker` isolates the disk and will
+not allocate slots on it until it becomes healthy again.
+
+Similarly, if usable space exceeds threshold (defaults to 5GiB), `Worker` will not allocate slots on it. In addition,
```
Comment (pan3793) [inline docs/developers/faulttolerant.md:63]: exceeds ?

**Comment by waitinfuture** [comment]
cc @cfmcgrady @pan3793 comments addressed, please take a look again

File: docs/developers/faulttolerant.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+Upon `PushMergedData` failure, `ShuffleClient` first unpacks and revives for every data batch. Notice that previously
+all data batches in `PushMergedData` have the same primary and replica (if any) destination, after reviving new
+`PartitionLocation`s can spread across multiple `Worker`s.
+
+Then `ShuffleClient` groups the new `PartitionLocations` in the same way as before, resulting in multiple
+`PushMergedData` requests, then send them to their destinations.
+
+Celeborn detects data lost when processing `CommitFiles` (See [Worker](../..developers/overview#shuffle-lifecycle)).
+Celeborn considers no `DataLost` IFF every `PartitionLocation` has succeeded to commit at least one replica (if
```
Comment (cfmcgrady) [inline docs/developers/faulttolerant.md:53]: IFF -> if and only if

Comment (waitinfuture) [reply docs/developers/faulttolerant.md]: done

**Review by cfmcgrady — APPROVED** [review]

**Review by pan3793 — APPROVED** [review]

**Comment by waitinfuture** [comment]
Thanks, merging to main

**Pull request closed without merging**